doc/lsst.daf.butler/dimensions.rst

```diff
@@ -0,0 +1,83 @@
+
+Dimension Relationships and Containers
+--------------------------------------
+
+Dimensions have can have relationships, and in fact these relationships are used almost exclusively to define relationships between datasets, which have no direct relationships between them.
+There are two kinds of relationships:
+
+ - If dimension ``A`` is a *required* dependency of dimension ``B``, ``A``'s primary key value must be provided in order to uniquely identify ``B``, and ``A`` has a one-to-many relationship with ``B``.
+   For example, the detector dimension has a required dependency on the instrument dimension, and hence one uses both an instrument name and a detector ID (or detector name) to fully identify a detector.
+   When both dimensions are associated with database tables, a required dependency involves having the dependency's primary key fields both foreign key fields and part of a compound primary key in the dependent table.
+
+ - If dimension ``C`` is an *implied* dependency of dimension ``D``, a value for ``D`` implies a value for ``C``, and ``C`` has a one-to-many relationship with ``D``.
+   For example, the visit dimension has an implied dependency on the physical filter dimension, because a visit is observed through exactly one filter and hence each visit ID determines a filter name.
+   When both dimensions are associated with database tables, an implied dependency involves having a foreign key field in the dependent table that is not a primary key in the dependent table.
+
+A `DimensionGraph` is an immutable, set-like container of dimensions that is guaranteed to (recursively) include all dependencies of any dimension in the graph.
+It also categorizes those dimensions into `~DimensionGraph.required` and `~DimensionGraph.implied` subsets, which have roughly the same meaning for a set of dimensions as they do for a single dimension: once the primary key values of all of the required dimensions are known, the primary key values of all implied dimensions are known as well.
+`DimensionGraph` also guarantees a deterministic and topological sort order for its elements.
+
+Because `Dimension` instances have a `~Dimension.name` attribute, we typically
+use `NamedValueSet` and `NamedKeyDict` as containers when immutability is needed or the guarantees of `DimensionGraph`.
+This allows the string names of dimensions to be used as well in most places where `Dimension` instances are expected.
+
+The complete set of all compatible dimensions is held by a special subclass of `DimensionGraph`, `DimensionUniverse`.
+A dimension universe is constructed from configuration, and is responsible for constructing all `Dimension` and `DimensionElement` instances; within a universe, there is exactly one `Dimension` instance that is always used to represent a particular dimension.
+
+`DimensionUniverse` instances themselves are held in a global map keyed by the version number in the configuration used to construction, so they behave somewhat like singletons.
```

Comment:
I haven't tried at all to come up with conventions or guidelines for the version numbers; all we have right now is code that blindly uses the version provided in the configuration as a key (which happens to be `zero`, since that's what's in the default configs).  And it's fine right now for us to not bother with meaningful version numbers, because we're not promising any kind of backwards compatibility yet, and the only reason we'll need versions (at least in the context described above) is so you can have different `Registry` instances with different dimensions alive in the same Python process, presumably because you're trying to utilize an older repository and a new one together for some reason.